src/swoop/api/routers/payloads.py

```diff
@@ -22,16 +50,58 @@
     response_model=PayloadList,
     responses={"404": {"model": APIException}},
 )
-def list_payloads(
-    limit: int = Query(ge=1, default=DEFAULT_JOB_LIMIT),
+async def list_payloads(
+    request: Request,
+    limit: int = Query(ge=1, default=DEFAULT_PAYLOAD_LIMIT),
     process_id: list[str] | None = Query(default=None),
     collection_id: list[str] | None = Query(default=None),
     item_id: list[str] | None = Query(default=None),
 ) -> PayloadList | APIException:
     """
     retrieve the list of payloads.
     """
-    pass
+    proc_clause = V("c.workflow_name") == funcs.any(process_id)
+    coll_clause = V("i.collection") == funcs.any(collection_id)
+    item_clause = V("i.item_id") == funcs.any(item_id)
+
+    async with request.app.state.readpool.acquire() as conn:
+        q, p = render(
+            """
+                SELECT DISTINCT(c.payload_uuid)
+                FROM swoop.item_payload i
+                INNER JOIN swoop.payload_cache c
+                ON c.payload_uuid = i.payload_uuid
+                INNER JOIN swoop.action a
+                ON c.payload_uuid = a.payload_uuid
+                WHERE (:processes::text[] IS NULL OR :proc_where)
+                AND (:collections::text[] IS NULL OR :coll_where)
+                AND (:items::text[] IS NULL OR :item_where)
```

Comment:
Can you help me understand why you have to cast to a text array here? It seems like the array is simply empty, which would not be `NULL` and the first condition could never be true. Or, if the array isn't an array and is `NULL`, then you shouldn't need to cast it to an array for the comparison.

But maybe I am missing something.

Reply:
Omitting the `::text[]` results in an `asyncpg.exceptions.IndeterminateDatatypeError`.  If no input is provided for `process_id`, `collection_id`, and/or `item_id`, we want to return all the payloads, whereas if an input is provided for these parameters, we want to filter the results so they match those parameters, correct?

The current behavior is that if no input is provided for one (or more) of these parameters, those parameters are of type None within the function handling the endpoint and the first clause (IS NULL) is true. If an input is provided for one (or more) of these parameters, the first clause is not true and the second clause (OR :proc_where, for example) applies.

Reply:
Thanks for all those details, that's helpful.

I do wonder if rather than casting the whole array to a type (the computation expense of which I am uncertain) if it would be better to ensure these parameters are always an array (i.e., never `NULL`), and simply check if they are empty in place of the comparison to `NULL`.

All that said, seeing as the values here are coming from url parameters, I suppose the length of any of these arrays is fairly well bounded by max url length restrictions. Therefore what I proposed is likely just premature optimization and leaving this alone is probably the right answer.